File: src/main/java/com/rakesh/assignment3/employee/EmployeeDataBase.java
package com.rakesh.assignment3.employee;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.ResultSet;
import java.sql.Statement;
import java.time.LocalDate;
import java.time.Period;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by Rakesh on Feb 16, 2022.
 */

public class EmployeeDataBase {
    static Connection con  = null;
    static Statement  stmt = null;
    static ResultSet  rst  = null;

    public static void getConnection() {
        try {
            Class.forName("org.postgresql.Driver");
            String url = "jdbc:postgresql://localhost:5432/training?user=postgres&password=admin";
            con = DriverManager.getConnection(url);
            stmt = con.createStatement();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private static void closeConnection() {
        try {
            stmt.close();
            con.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static List<Employee1> getAllEmployees() {
        List<Employee1> result = new ArrayList<>();
        getConnection();
        try {
            rst = stmt.executeQuery("Select * from public.\"Employee\"");
            while (rst.next()) {
                LocalDate joiningDate = LocalDate.parse(rst.getString("JoingDate"));
                LocalDate dob         = LocalDate.parse(rst.getString("DOB"));
                result.add(new Employee1(rst.getInt("EmpID"), rst.getString("Name"), rst.getString("Department")
                        , rst.getDouble("Salary"), rst.getString("Gender"), joiningDate, dob, rst.getString("JobLevel")));
            }
        } catch (Exception e) {
            System.out.println(e.getMessage());
        }
        closeConnection();
        return result;
    }

    public static Employee1 getEmployeeByID(int empID) {
        Employee1 result = null;
        getConnection();
        try {
            rst = stmt.executeQuery("Select * from public.\"Employee\" where \"EmpID\"=" + empID);
            if (rst.next()) {
                LocalDate joiningDate = LocalDate.parse(rst.getString("JoingDate"));
                LocalDate dob         = LocalDate.parse(rst.getString("DOB"));
                result = new Employee1(rst.getInt("EmpID"), rst.getString("Name"), rst.getString("Department")
                        , rst.getDouble("Salary"), rst.getString("Gender"), joiningDate, dob, rst.getString("JobLevel"));
            }
        } catch (Exception e) {
            System.out.println(e.getMessage());
        }
        closeConnection();
        return result;
    }

    public static void addEmployee(Employee1 employee) {
        getConnection();
        int age = Period.between(employee.getDob(), LocalDate.now()).getYears();
        if (age >= 21) {
            DateTimeFormatter dtf       = DateTimeFormatter.ofPattern("yyyy-MM-dd");
            String            joingDate = dtf.format(employee.getJoiningDate());
            String            dob       = dtf.format(employee.getDob());
            try {
                stmt.executeUpdate("insert into public.\"Employee\"(\"EmpID\",\"Name\",\"Department\",\"Salary\",\"Gender\",\"JoingDate\",\"DOB\",\"JobLevel\")" +
                        " values(" + employee.getEmpID() + ",\'" + employee.getName() + "\',\'" + employee.getDepartment() + "\'," + employee.getSalary() + "," +
                        "\'" + employee.getGender() + "\',\'" + joingDate + "\',\'" + dob + "\',\'" + employee.getJobLevel() + "\')");

            } catch (Exception e) {
                System.out.println(e.getMessage());
            }
            closeConnection();
        } else {
            throw new RuntimeException("Employee age is under 21 years");
        }
    }

    public static void updateData(String columnName, String newValue, int empID) {
        getConnection();
        if (columnName.equalsIgnoreCase("Name")) {
            try {
                stmt.executeUpdate("update public.\"Employee\" set \"Name\" =\'" + newValue + "\' where \"EmpID\"=" + empID);
            } catch (Exception e) {
                System.err.println(e.getMessage());
            }
        } else if (columnName.equalsIgnoreCase("Department")) {
            try {
                stmt.executeUpdate("update public.\"Employee\" set \"Department\" =\'" + newValue + "\' where \"EmpID\"=" + empID);
            } catch (Exception e) {
                System.err.println(e.getMessage());
            }
        } else if (columnName.equalsIgnoreCase("salary")) {
            try {
                double salary = Double.parseDouble(newValue);
                stmt.executeUpdate("update public.\"Employee\" set \"Salary\" =\'" + salary + "\' where \"EmpID\"=" + empID);
            } catch (Exception e) {
                System.err.println(e.getMessage());
            }
        } else if (columnName.equalsIgnoreCase("Gender")) {
            try {
                stmt.executeUpdate("update public.\"Employee\" set \"Gender\" =\'" + newValue + "\' where \"EmpID\"=" + empID);
            } catch (Exception e) {
                System.err.println(e.getMessage());
            }
        } else if (columnName.equalsIgnoreCase("joiningDate")) {
            try {
                DateTimeFormatter dtf       = DateTimeFormatter.ofPattern("yyyy-MM-dd");
                LocalDate         joingdate = LocalDate.parse(newValue, dtf);
                int               days      = Period.between(joingdate, LocalDate.now()).getDays();
                if (days > -1) {
                    stmt.executeUpdate("update public.\"Employee\" set \"JoingDate\" =\'" + newValue + "\' where \"EmpID\"=" + empID);
                } else {
                    throw new RuntimeException("Invalid Joining date");
                }
            } catch (Exception e) {
                System.err.println(e.getMessage());
            }
        } else if (columnName.equalsIgnoreCase("DOB")) {
            try {
                DateTimeFormatter dtf = DateTimeFormatter.ofPattern("yyyy-MM-dd");
                LocalDate         dob = LocalDate.parse(newValue, dtf);
                int               age = Period.between(dob, LocalDate.now()).getYears();
                if (age >= 21) {
                    stmt.executeUpdate("update public.\"Employee\" set \"DOB\" =\'" + newValue + "\' where \"EmpID\"=" + empID);
                } else {
                    throw new RuntimeException("Employee age cannot be under 21");
                }
            } catch (Exception e) {
                System.err.println(e.getMessage());
            }
        } else if (columnName.equalsIgnoreCase("jobLevel")) {
            try {
                stmt.executeUpdate("update public.\"Employee\" set \"JobLevel\" =\'" + newValue + "\' where \"EmpID\"=" + empID);
            } catch (Exception e) {
                System.err.println(e.getMessage());
            }
        } else {
            throw new RuntimeException("No Column Name found in the DataBase");
        }
        closeConnection();
    }

    public static void deleteEmployeeByID(int empID) {
        getConnection();
        try {
            stmt.execute("Delete from public.\"Employee\" where \"EmpID\"=" + empID);
        } catch (Exception e) {
            System.err.println(e.getMessage());
        }
        closeConnection();
    }

    public static void deleteAllEmployee() {
        getConnection();
        try {
            stmt.executeUpdate("Delete from public.\"Employee\"");
        } catch (Exception e) {
            e.printStackTrace();
        }
        closeConnection();
    }

//    public static void main(String[] args) {
//       // deleteEmployeeByID(1005);
//       // deleteAllEmployee();
//        updateData("Joblevel", "Junior", 1004);
//    }
}

File: src/main/scala/com/chethan/assignment1/EmployeeDatabase.scala
package com.chethan.assignment1

import java.sql._
import java.time.format.DateTimeFormatter
import java.time.{LocalDate, Period}

import scala.concurrent._
import scala.util.{Failure, Success, Try}


/**
  * Created by Chethan on Jun 13, 2022.
  */

class EmployeeDatabase extends DataBase[Employee] {

  private def getConnection(host: String, username: String, password: String): Connection = {
    val driver  = "org.postgresql.Driver"
    val jdbcUrl = s"jdbc:postgresql://$host:5432/training?user=$username&password=$password"
    Class forName driver
    DriverManager getConnection jdbcUrl
  }

  override def getAllData: Future[List[Employee]] = {
    val getAllQuery = s""" SELECT * FROM "public"."Employee" """
    val format      = DateTimeFormatter ISO_LOCAL_DATE

    def getAllFromResultSet(resultSet: ResultSet): List[Employee] = {
      def resultSetLoop(resultSetData: ResultSet, listOfEmployees: List[Employee]): List[Employee] = {
        if (!resultSet.next())
          listOfEmployees
        else {
          val id          = resultSet getInt "id"
          val name        = resultSet getString 2
          val department  = resultSet getString 3
          val salary      = resultSet getDouble 4
          val gender      = resultSet getString 5
          val joiningDate = LocalDate parse(resultSet getString 6, format)
          val dob         = LocalDate parse(resultSet getString 7, format)
          val jobLevel    = resultSet getString 8
          resultSetLoop(resultSetData, Employee(id, name, department, salary, gender, joiningDate, dob, jobLevel) :: listOfEmployees)
        }
      }

      resultSetLoop(resultSet, Nil)
    }

    val result = Try[List[Employee]]{
      val connection      = new EmployeeDatabase() getConnection("127.0.0.1", "postgres", "admin")
      val statement       = connection createStatement()
      val resultSet       = statement executeQuery getAllQuery
      val listOfEmployees = getAllFromResultSet(resultSet)
      resultSet close()
      statement close()
      connection close()
      listOfEmployees
    }

    result match {
      case Success(value)     => Future successful value
      case Failure(exception) => Future failed exception
    }
  }

  override def getDataById(id: Any): Future[Employee] = {
    val format           = DateTimeFormatter ISO_LOCAL_DATE
    val getDataByIdQuery = s"""SELECT * FROM "public"."Employee" WHERE id = $id """

    val result = Try[Employee]{
      val connection = new EmployeeDatabase() getConnection("127.0.0.1", "postgres", "admin")
      val statement  = connection createStatement()
      val resultSet  = statement executeQuery getDataByIdQuery
      resultSet next()
      val dataId      = resultSet getInt "id"
      val name        = resultSet getString 2
      val department  = resultSet getString 3
      val salary      = resultSet getDouble 4
      val gender      = resultSet getString 5
      val joiningDate = LocalDate parse(resultSet getString 6, format)
      val dob         = LocalDate parse(resultSet getString 7, format)
      val jobLevel    = resultSet getString 8
      resultSet close()
      statement close()
      connection close()
      Employee(dataId, name, department, salary, gender, joiningDate, dob, jobLevel)
    }
    result match {
      case Success(value)     => Future successful value
      case Failure(exception) => Future failed exception
    }
  }

  override def addData(employee: Employee): Future[Boolean] = {
    val getInsertQuery =
      s""" INSERT INTO "public"."Employee" (id, name, department, salary, gender, \"joiningDate\", dob, \"jobLevel\") values ('${employee.id}','${employee.name}','${employee.department}','${employee.salary}','${employee.gender}','${employee.joiningDate}','${employee.dob}','${employee.jobLevel}')"""

    val result = Try[Boolean]{
      val connection = new EmployeeDatabase() getConnection("127.0.0.1", "postgres", "admin")
      val statement  = connection createStatement()
      val isSuccess  = if (Period.between(employee.dob, LocalDate.now).getYears >= 21 && !Period.between(employee.joiningDate, LocalDate.now).isNegative && Period.between(employee.joiningDate, employee.dob).isNegative) {
        statement executeUpdate getInsertQuery
        true
      } else throw new RuntimeException
      statement close()
      connection close()
      isSuccess
    }
    result match {
      case Success(value)     => Future successful value
      case Failure(exception) => Future failed exception
    }
  }

  override def updateData(employee: Employee): Future[Boolean] = {
    val getInsertQuery =
      s""" UPDATE "public"."Employee" SET id = '${employee.id}', name = '${employee.name}' , department = '${employee.department}' , salary = '${employee.salary}', gender = '${employee.gender}', \"joiningDate\" = '${employee.joiningDate}',dob = '${employee.dob}', \"jobLevel\" = '${employee.jobLevel}' WHERE id = '${employee.id}'"""
    val result         = Try[Boolean]{
      val connection = new EmployeeDatabase() getConnection("127.0.0.1", "postgres", "admin")
      val statement  = connection createStatement()
      val isSuccess  = if (Period.between(employee.dob, LocalDate.now).getYears >= 21 && !Period.between(employee.joiningDate, LocalDate.now).isNegative && Period.between(employee.joiningDate, employee.dob).isNegative) {
        statement executeUpdate getInsertQuery
        true
      } else false
      statement close()
      connection close()
      isSuccess
    }

    result match {
      case Success(value)     => Future successful value
      case Failure(exception) => Future failed exception
    }
  }

  override def deleteData(id: Any): Future[Boolean] = {
    val getInsertQuery = s""" DELETE FROM "public"."Employee" WHERE id = '$id'"""
    val result         = Try[Boolean]{
      val connection = new EmployeeDatabase() getConnection("127.0.0.1", "postgres", "admin")
      val statement  = connection createStatement()
      statement executeUpdate getInsertQuery
      statement close()
      connection close()
      true
    }
    result match {
      case Success(value)     => Future successful value
      case Failure(exception) => Future failed exception
    }
  }
}

case class Employee(id: Int, name: String, department: String, salary: Double, gender: String, joiningDate: LocalDate, dob: LocalDate, jobLevel: String)




